crates/dids/src/method/dht/mod.rs

```diff
@@ -0,0 +1,148 @@
+mod bep44;
+mod convert;
+mod registered_types;
+
+use std::sync::Arc;
+
+use crypto::Curve;
+use keys::key_manager::KeyManager;
+use reqwest::blocking::Client;
+
+use crate::{
+    bearer::BearerDid,
+    document::{Document, Service, VerificationMethod},
+    identifier::Identifier,
+    resolver::{ResolutionMetadata, ResolutionResult},
+};
+
+use self::{
+    bep44::encode_bep44_message, convert::document_packet::document_to_packet,
+    registered_types::RegisteredDidType,
+};
+
+use super::{Method, MethodError};
+
+const DEFAULT_TTL: u32 = 7200; // seconds
+const DEFAULT_GATEWAY_URL: &str = "https://diddht.tbddev.org";
+const DID_DHT_SPEC_VERSION: u32 = 0;
+const NAME: &'static str = "dht";
+
+/// Concrete implementation for a did:dht DID
+pub struct DidDht;
+
+/// Options that can be used to create a did:dht DID
+pub struct DidDhtCreateOptions {
+    pub publish: bool,
+    // TODO: gatewayUri - default 'https://diddht.tbddev.org'
+    pub also_known_as: Option<Vec<String>>,
+    pub controller: Option<Vec<String>>,
+    pub service: Option<Vec<Service>>,
+    pub registered_type: Option<Vec<RegisteredDidType>>,
+    pub verification_methods: Option<Vec<VerificationMethod>>,
+}
+
+impl Method<DidDhtCreateOptions> for DidDht {
+    const NAME: &'static str = "dht";
+
+    fn create(
+        key_manager: Arc<dyn KeyManager>,
+        options: DidDhtCreateOptions,
+    ) -> Result<BearerDid, MethodError> {
+        // Generate private and public keypair
+        let key_alias = key_manager.generate_private_key(
+            Curve::Ed25519,
+            Some("0".to_string()), // PR review: i copied this from jwk.rs. do we want this to be None?
```

Comment:
this is correct BTW, and this is a new paradigm only made available in rust

which is, the `KeyManager` trait's `generate_private_key` enables the caller to override setting the `key_alias` (in the case of did:dht and did:jwk, both have an ID of `0`), which is necessary for subsequently instantiating a bearer DID from an existing key manager in order for the Verification Method's key ID to match the key alias in the key manager

Reply:
this is made possible in other SDKs through the portable DID, but I'm making it an initiative to second-class the serialization of private key material and instead make secure enclave the first class assumption (which is to say, keys are never serialized)